src/coap_mbedtls.c

```diff
@@ -1270,8 +1273,134 @@ mbedtls_debug_out(void *ctx COAP_UNUSED, int level,
   coap_log(log_level, "%s:%04d: %s", file, line, str);
 }
 
+#if !COAP_DISABLE_TCP
+/*
+ * return +ve data amount
+ *        0   no more
+ *        -ve  Mbed TLS error
+ */
+static int
+coap_sock_read(void *ctx, unsigned char *out, size_t outl) {
+  int ret = MBEDTLS_ERR_SSL_CONN_EOF;
+  coap_session_t *c_session = (coap_session_t *)ctx;
+
+  if (out != NULL) {
+#ifdef _WIN32
+    ret = recv(c_session->sock.fd, (char *)out, (int)outl, 0);
+#else
+    ret = recv(c_session->sock.fd, out, outl, 0);
+#endif
+    if (ret > 0) {
+      coap_log(LOG_DEBUG, "*  %s: received %d bytes\n",
+               coap_session_str(c_session), ret);
+    } else if (ret < 0 && errno != EAGAIN) {
+      coap_log(LOG_DEBUG,  "*  %s: failed to receive any bytes (%s)\n",
+               coap_session_str(c_session), coap_socket_strerror());
+    }
+
+    if (ret == 0) {
+      /* graceful shutdown */
+      c_session->sock.flags &= ~COAP_SOCKET_CAN_READ;
+      ret = MBEDTLS_ERR_SSL_CONN_EOF;
+    }
+    else if (ret == COAP_SOCKET_ERROR) {
+#ifdef _WIN32
+      int lasterror = WSAGetLastError();
+
+      if (lasterror == WSAEWOULDBLOCK) {
+        ret = MBEDTLS_ERR_SSL_WANT_READ;
+      }
+      else if (lasterror == WSAECONNRESET) {
+        ret = MBEDTLS_ERR_NET_CONN_RESET;
+      }
+#else
+      if (errno == EAGAIN) {
+        ret = MBEDTLS_ERR_SSL_WANT_READ;
+      }
+      else if (errno == EPIPE || errno == ECONNRESET) {
+        ret = MBEDTLS_ERR_NET_CONN_RESET;
+      }
+#endif
+      else {
+        ret = MBEDTLS_ERR_NET_RECV_FAILED;
+      }
+      c_session->sock.flags &= ~COAP_SOCKET_CAN_READ;
+    }
+    else if (ret < (ssize_t)outl) {
+      c_session->sock.flags &= ~COAP_SOCKET_CAN_READ;
+    }
+  }
+  return ret;
+}
+
+/*
+ * return +ve data amount
+ *        0   no more
+ *        -ve  Mbed TLS error
+ */
+static int
+coap_sock_write(void *context, const unsigned char *in, size_t inl) {
+  int ret = 0;
+  coap_session_t *c_session = (coap_session_t *)context;
+
+  ret = (int)coap_socket_write(&c_session->sock, in, inl);
```

Comment:
I suggest using `ssize_t` in this function and do the cast to `int` for the return only. (But this is more a question of taste and personal preferences, I suppose.)

Reply:
CHange made as suggested.  Code pushed.